src/platform/Linux/bluez/ChipDeviceScanner.cpp

```diff
@@ -110,33 +97,28 @@ ChipDeviceScanner::~ChipDeviceScanner()
         chip::DeviceLayer::SystemLayer().CancelTimer(TimerExpiredCallback, this);
     }
 
-    g_object_unref(mManager);
-    g_object_unref(mCancellable);
-    g_object_unref(mAdapter);
-
-    mManager     = nullptr;
-    mAdapter     = nullptr;
-    mCancellable = nullptr;
-    mDelegate    = nullptr;
-}
-
-std::unique_ptr<ChipDeviceScanner> ChipDeviceScanner::Create(BluezAdapter1 * adapter, ChipDeviceScannerDelegate * delegate)
-{
-    GDBusCreateObjectManagerContext context;
-    CHIP_ERROR err;
-
-    err = PlatformMgrImpl().GLibMatterContextInvokeSync(MainLoopCreateObjectManager, &context);
-    VerifyOrExit(err == CHIP_NO_ERROR, ChipLogError(Ble, "Failed to create BLE object manager"));
-
-    return std::make_unique<ChipDeviceScanner>(context.object, adapter, context.cancellable, delegate);
-
-exit:
-    return std::unique_ptr<ChipDeviceScanner>();
+    // Release resources on the glib thread. This is necessary because the D-Bus manager client
+    // object handles D-Bus signals. Otherwise, we might face a race when the manager object is
+    // released during a D-Bus signal being processed.
+    PlatformMgrImpl().GLibMatterContextInvokeSync(
+        +[](ChipDeviceScanner * self) {
+            if (self->mManager != nullptr)
+                g_object_unref(self->mManager);
+            if (self->mAdapter != nullptr)
+                g_object_unref(self->mAdapter);
+            if (self->mCancellable != nullptr)
+                g_object_unref(self->mCancellable);
+            return CHIP_NO_ERROR;
+        },
+        this);
+
+    mIsInitialized = false;
 }
 
 CHIP_ERROR ChipDeviceScanner::StartScan(System::Clock::Timeout timeout)
 {
     assertChipStackLockedByCurrentThread();
+    ReturnErrorCodeIf(mIsInitialized, CHIP_ERROR_INCORRECT_STATE);
```

Comment:
It should be here `ReturnErrorCodeIf(!mIsInitialized, CHIP_ERROR_INCORRECT_STATE);`

Reply:
Many thanks for review! Fixed.